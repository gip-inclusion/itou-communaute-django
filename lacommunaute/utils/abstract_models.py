from django.conf import settings
from django.db import models
from django.utils.encoding import force_str
from django.utils.text import slugify
from django.utils.translation import gettext_lazy as _
from machina.models.fields import MarkupTextField
from storages.backends.s3boto3 import S3Boto3Storage

from lacommunaute.utils.validators import validate_image_size


class DatedModel(models.Model):
    created = models.DateTimeField(auto_now_add=True, verbose_name=_("Creation date"))
    updated = models.DateTimeField(auto_now=True, verbose_name=_("Update date"))

    class Meta:
        abstract = True


class Publication(DatedModel):
    name = models.CharField(max_length=100, verbose_name=_("Name"), unique=True)
    slug = models.SlugField(max_length=255, verbose_name=_("Slug"))

    description = MarkupTextField(verbose_name=_("Description"), null=True, blank=True)
    short_description = models.CharField(max_length=400, verbose_name=_("Short Description"))
    image = models.ImageField(
        storage=S3Boto3Storage(bucket_name=settings.AWS_STORAGE_BUCKET_NAME, file_overwrite=False),
        validators=[validate_image_size],
        blank=True,
        null=True,
    )

    class Meta:
        abstract = True

    def __str__(self):
        return f"{self.name}"

    def save(self, *args, **kwargs):
        self.slug = slugify(force_str(self.name), allow_unicode=True)
        super().save(*args, **kwargs)
