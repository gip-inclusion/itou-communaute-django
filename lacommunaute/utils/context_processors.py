from urllib.parse import urlencode

from django.conf import settings


def expose_settings(request):
    """
    Put things into the context to make them available in templates.
    https://docs.djangoproject.com/en/4.1/ref/templates/api/#using-requestcontext
    """
    base_template = "layouts/base_htmx.html" if getattr(request, "htmx", False) else "layouts/base.html"

    return {
        "BASE_TEMPLATE": base_template,
        "MATOMO_SITE_ID": settings.MATOMO_SITE_ID,
        "MATOMO_BASE_URL": settings.MATOMO_BASE_URL,
        "ENVIRONMENT": settings.ENVIRONMENT,
        "EMPLOIS_PRESCRIBER_SEARCH": settings.EMPLOIS_PRESCRIBER_SEARCH,
        "EMPLOIS_COMPANY_SEARCH": settings.EMPLOIS_COMPANY_SEARCH,
    }



def matomo(request):
    if not request.resolver_match:
        return {"send_to_matomo": False}

    context = {"send_to_matomo": True}
    url = request.resolver_match.route
    # Only keep Matomo-related params for now.
    params = {k: v for k, v in request.GET.lists() if k.startswith(("utm_", "mtm_", "piwik_"))}
    if params:
        url = f"{url}?{urlencode(sorted(params.items()), doseq=True)}"
    context["matomo_custom_url"] = url
    return context
