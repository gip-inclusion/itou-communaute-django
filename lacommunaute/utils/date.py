from datetime import date

from dateutil.relativedelta import relativedelta


def get_last_sunday(theday=date.today()):
    days_to_subtract = (theday.weekday() + 1) % 7
    return theday - relativedelta(days=days_to_subtract)
