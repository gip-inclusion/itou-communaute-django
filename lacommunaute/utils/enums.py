from django.db import models


class PeriodAggregation(models.TextChoices):
    MONTH = "MONTH"
    WEEK = "WEEK"
