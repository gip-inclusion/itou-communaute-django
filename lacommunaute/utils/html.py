import re


def wrap_iframe_in_div_tag(text):
    """
    given a markdown text, wrap all iframe tags in a div tag
    this is required for iframes to be displayed correctly
    """

    iframe_regex = r"((<div>)?(<iframe[^>]*><\/iframe>)(<\/div>)?)"

    for match, starts_with, iframe, ends_with in re.findall(iframe_regex, text):
        if not (starts_with and ends_with):
            text = text.replace(match, f"<div>{iframe}</div>")

    return text
