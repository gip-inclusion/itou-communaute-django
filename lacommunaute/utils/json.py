# temp file. To be removed after datas are in the database

import json


def get_json_data(file_name):
    try:
        with open(file_name) as json_file:
            data = json.load(json_file)
            return data
    except FileNotFoundError:
        return []


def extract_values_in_list(datas):
    values = {
        "date": [],
        "nb_unique_contributors": [],
        "nb_uniq_visitors": [],
        "nb_uniq_active_visitors": [],
        "nb_engagment_events": [],
    }
    for data in datas:
        values[data["name"]].append(data["value"])
        if data["date"] not in values["date"]:
            values["date"].append(data["date"])
    return values
