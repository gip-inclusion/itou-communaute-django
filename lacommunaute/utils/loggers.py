from json_log_formatter import JSONFormatter


class CustomJsonFormatter(JSONFormatter):
    def json_record(self, message, extra, record):
        extra["logger_name"] = record.name

        if "context" in extra:
            context = extra.pop("context")
            if hasattr(context, "request"):
                extra = extra | {
                    "view": context.request.resolver_match.view_name,
                    "kwargs": context.request.resolver_match.kwargs,
                    "method": context.request.method,
                    "user_id": context.request.user.id if context.request.user.is_authenticated else None,
                    "session_key": context.request.session.session_key,
                }

        return super().json_record(message, extra, record)
