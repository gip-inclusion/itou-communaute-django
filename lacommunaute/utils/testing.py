import importlib

from bs4 import BeautifulSoup
from django.db import connection
from django.test.utils import TestContextDecorator


class reload_module(TestContextDecorator):
    def __init__(self, module):
        self._module = module
        self._original_values = {key: getattr(module, key) for key in dir(module) if not key.startswith("__")}
        super().__init__()

    def enable(self):
        importlib.reload(self._module)

    def disable(self):
        for key, value in self._original_values.items():
            setattr(self._module, key, value)


def parse_response_to_soup(response, selector=None, no_html_body=False, replace_in_href=None):
    soup = BeautifulSoup(response.content, "html5lib", from_encoding=response.charset or "utf-8")
    if no_html_body:
        # If the provided HTML does not contain <html><body> tags
        # html5lib will always add them around the response:
        # ignore them
        soup = soup.body
    if selector is not None:
        [soup] = soup.select(selector)
    for csrf_token_input in soup.find_all("input", attrs={"name": "csrfmiddlewaretoken"}):
        csrf_token_input["value"] = "NORMALIZED_CSRF_TOKEN"
    if "nonce" in soup.attrs:
        soup["nonce"] = "NORMALIZED_CSP_NONCE"
    for csp_nonce_script in soup.find_all("script", {"nonce": True}):
        csp_nonce_script["nonce"] = "NORMALIZED_CSP_NONCE"
    if replace_in_href:
        replacements = [
            (
                replacement
                if isinstance(replacement, tuple)
                else (str(replacement.pk), f"[PK of {type(replacement).__name__}]")
            )
            for replacement in replace_in_href
        ]
        for attr in ["href", "hx-post"]:
            for links in soup.find_all(attrs={attr: True}):
                [links.attrs.update({attr: links.attrs[attr].replace(*replacement)}) for replacement in replacements]
    return soup


def reset_model_sequence_fixture(model_class):
    """
    :return: a function which can adjust and reset a primary key sequence for use as a pytest fixture
    it is used to temporarily change the primary key, so that it is predictable (e.g. for snapshots)
    """

    def reset_model_sequence():
        prior_seq_val = 1

        def get_current_sequence_value(cursor):
            cursor.execute(f"SELECT nextval(pg_get_serial_sequence('{model_class._meta.db_table}', 'id'));")
            return cursor.fetchone()[0]

        def set_sequence_value(cursor, value):
            cursor.execute(
                f"SELECT setval(pg_get_serial_sequence('{model_class._meta.db_table}', 'id'), '{str(value)}', false);"
            )

        with connection.cursor() as cursor:
            prior_seq_val = get_current_sequence_value(cursor)
            set_sequence_value(cursor, 9999)

        yield

        with connection.cursor() as cursor:
            set_sequence_value(cursor, prior_seq_val)
            assert prior_seq_val == get_current_sequence_value(cursor)

    return reset_model_sequence
