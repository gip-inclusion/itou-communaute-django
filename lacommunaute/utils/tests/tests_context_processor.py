import pytest
from django.test import override_settings

from lacommunaute.utils.enums import Environment
from lacommunaute.utils.testing import parse_response_to_soup


@pytest.mark.parametrize(
    "env,expected",
    [
        (None, False),
        (Environment.PROD, False),
        (Environment.TEST, True),
        (Environment.DEV, True),
    ],
)
def test_prod_environment(client, db, env, expected, snapshot):
    with override_settings(ENVIRONMENT=env):
        response = client.get("/")
    assert ('id="debug-mode-banner"' in response.content.decode()) == expected
    if expected:
        content = parse_response_to_soup(response, selector="#debug-mode-banner")
        assert str(content) == snapshot(name=env.label)


def test_exposed_settings(client, db):
    response = client.get("/")
    assert "BASE_TEMPLATE" in response.context
    assert "MATOMO_SITE_ID" in response.context
    assert "MATOMO_BASE_URL" in response.context
    assert "ENVIRONMENT" in response.context
    assert "EMPLOIS_PRESCRIBER_SEARCH" in response.context
    assert "EMPLOIS_COMPANY_SEARCH" in response.context
