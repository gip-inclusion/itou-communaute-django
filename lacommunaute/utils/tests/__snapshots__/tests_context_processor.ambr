# serializer version: 1
# name: test_prod_environment[DEV-True][Dev]
  '''
  <div class="global-messages-container" id="debug-mode-banner">
                      <div class="alert alert-danger fade show" role="status">
                          <p class="mb-0">
                              <strong class="text-uppercase">DEV MODE</strong>
                          </p>
                      </div>
                  </div>
  '''
# ---
# name: test_prod_environment[TEST-True][Test]
  '''
  <div class="global-messages-container" id="debug-mode-banner">
                      <div class="alert alert-danger fade show" role="status">
                          <p class="mb-0">
                              <strong class="text-uppercase">TEST MODE</strong>
                          </p>
                      </div>
                  </div>
  '''
# ---
